plugin/formatting.py

```diff
@@ -56,11 +67,74 @@ def run(self, _):
                     "uri": filename_to_uri(self.view.file_name())
                 },
                 "range": Range.from_region(self.view, region).to_lsp(),
-                "options": {
-                    "tabSize": self.view.settings().get("tab_size", 4),
-                    "insertSpaces": True
-                }
+                "options": get_formatting_options(self.view)
             }
             client.send_request(Request.rangeFormatting(params),
                                 lambda response: self.view.run_command('lsp_apply_document_edit',
                                                                        {'changes': response}))
+
+
+class DocumentOnTypeFormattingListener(sublime_plugin.ViewEventListener):
+
+    @classmethod
+    def is_applicable(cls, subl_settings: sublime.Settings) -> bool:
+        syntax = subl_settings.get('syntax')
+        if not syntax:
+            return False
+        elif not is_supported_syntax(syntax):
+            return False
+        return True
+
+    def __init__(self, view: sublime.View) -> None:
+        super().__init__(view)
+        self._triggers = []  # type: List[str]
+        self._client = None  # type: Optional[Client]
+        self._initialized = False
+
+    def on_modified_async(self) -> None:
+        if not settings.format_on_type:
+            return
+        if not self._ensure_initialized():
+            return
+        for region in self.view.sel():
+            self._on_type_formatting_async(region.begin())
+
+    def _ensure_initialized(self) -> bool:
+        if self._initialized:
+            return bool(self._client)
+        self._initialized = True
+        self._client = client_for_view(self.view)
+        if self._client:
+            options = self._client.get_capability("documentOnTypeFormattingProvider")
+            if options:
+                self._triggers = [options.get("firstTriggerCharacter")] or []
```

Comment:
I may be misunderstanding, but if "firstTriggerCharacter" is absent the result would be:
```python
# Python 2.7.12 (default, Jun 29 2016, 14:05:02)
# [GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)] on darwin
# Type "help", "copyright", "credits" or "license" for more information.
>>> [None] or []
[None]
```
Following the spec, "firstTriggerCharacter" must be present. I'd prefer to disable the feature if it's missing.

Reply:
done